# DEICODE
## Discovery of Environmental Influences through Convex Optimized Decomposition by Ecotypes (DEIECODE) 
## Still Under Development 

By decomposing the sparse data into its dense low-rank component trends can be discorved from enviornmental cgnages between sample groups.

This project is a demonstration of this method in 16S rRNA sequencing data. 

## Installation

cd usr../DEIECODE

chmod +x setup.sh

./setup.sh

## Usage

cd usr.../DEIECODE 

source activate COD_env 

jupyter notebook 

A browser window will open, then open notebooks to view examples from paper. Shift-enter to run cells. 


## Contributing

1. Fork it!
2. Create your feature branch: `git checkout -b my-new-feature`
3. Commit your changes: `git commit -am 'Add some feature'`
4. Push to the branch: `git push origin my-new-feature`
5. Submit a pull request :D

## Credits

Source decomposition code and algorithm is from https://github.com/dganguli/robust-pca

Soil and Human microbiome examples of use in biological data written by cameron martino 

for issues please contact cameronmartino at gmail dot com 

